### PR TITLE
More efficient template representation

### PR DESCRIPTION
**Pull request by jods4**
This is work in progress. It needs clean-up but it does work and passes all tests already.

Less allocations, better cache locality. No need for virtual calls during rendering.

### CODE REVIEW

**Comment by jods4** [comment]
Some parsing benchmarks in the current state:

For templates with almost no literal text, the parsing speed is the same as the old one (within one stddev):
`"{Timestamp:yyyy-MM-dd HH:mm:ss} [{Level}] {Message}{NewLine}{Exception}"`

| Method | Median | StdDev |
| --- | --- | --- |
| OldFormat | 699.5150 ns | 20.9761 ns |
| NewFormat | 713.9544 ns | 17.7361 ns |

For templates with more text, the fewer copies are very visible, 25% improvement on this one:
`"Line from pt1 {start} to pt2 {end} with length {length}."`

| Method | Median | StdDev |
| --- | --- | --- |
| OldFormat | 626.1473 ns | 16.8222 ns |
| NewFormat | 474.2144 ns | 13.5524 ns |

And of course, the better locality of data should be positive for rendering as well.

**Comment by 304NotModified** [comment]
AFAIK the recommendation for a struct is max 20 bytes. Are some structs not to large now?

**Comment by 304NotModified** [comment]
I like the approach! Great work. It makes rendering a bit more complex, but I think it's worth it.

**Comment by 304NotModified** [comment]
~~pushed a merge fix. I like rebasing more, but then you should use-- `git pull --rebase`!~~

File: src/Parser/Template.cs [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+        }
+      }
+      return sb.ToString();
+    }
+  }
+
+  public struct Literal
+  {
+    // Number of characters to print
+    public ushort Print;
```
Comment (304NotModified) [inline src/Parser/Template.cs]: nice, but isn't an issue that it isn't cls compliant?

File: src/Parser/TemplateParser.cs [hunk 2]
```diff
@@ -1,211 +1,233 @@
...
-            {
-                var formattedChars = string.Join(", ", search.Select(c => "'" + c + "'").ToArray()); 
-                throw new TemplateParserException($"Reached end of template while expecting one of {formattedChars}.", _position);
-            }
-            _position = i == -1 ? _length : i;
-            return _template.Substring(start, _position - start);
-        }
-    }
-}
+﻿using System;
```
Comment (304NotModified) [inline src/Parser/TemplateParser.cs]: The newlines is an issue I see, will fix that

**Comment by 304NotModified** [comment]
I did a `rebase` to fix the EOL issues like:

image

please use `git pull --rebase`. Sorry if this will gives you trouble.

File: src/Parser/Template.cs [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
```
Comment (304NotModified) [inline src/Parser/Template.cs]: Nice, but isn't an issue that this isn't CLS compliant?

File: src/Parser/TemplateParser.cs [hunk 4]
```diff
@@ -86,17 +91,28 @@ private void ParseCloseBracketPart()
...
             Skip('}');
-            _parts.Add(new HolePart(name, type, position, format, aligment));
+            _result.Literals.Add(new Literal { Print = _print, Skip = (ushort)(_position - start + (type == HoleType.Normal ? 1 : 2)) });
+            _print = 0;
+            _result.Holes.Add(new Hole 
+            { 
+                Name = name,
+                Format = format,
+                CaptureType = type,
+                Index = (byte)position,
```
Comment (304NotModified) [inline src/Parser/TemplateParser.cs:123]: Guess we will remove all of these casts in the near future?

**Comment by jods4** [comment]
> AFAIK the recommendation for a struct is max 20 bytes. Are some structs not to large now?

If I'm not mistaken, Hole struct is 12 bytes on 32bits, 20 on 64bits. Difference is caused by 2 ref pointers.
The recommendation is because struct are value types, hence entirely copied each time you pass them around which gets costly if your struct is large.
I don't plan to copy ours around much. For rendering I currently have one copy in mind, hopefully I can make it 0, but without ref-returns (C# 7) I'm not sure if it's possible.
The main goal here was to reduce memory allocations, and have all the holes and literals data sitting in contiguous memory.

>  It makes rendering a bit more complex

Not even that much. My `Rebuild` method is basically the skeleton for rendering as well. And I also think it's worth it if you want to maximize perf.

> nice, but isn't an issue that it isn't cls compliant?

To be honest, I am no expert about CLS-compliance as I never cared about this in my projects before.
But as far as I understand the topic: CLS-compliance is there to ensure your lib can be consumed in all .NET languages, not only C#/VB.NET. Only the public surface needs to be CLS-compliant, private implementation details are fine.

I think those struct are internal implementation, right? Non-structured targets will receive a string; Structured targets will receive the template string + a list of key-value pairs. 

So this is a non-issue, isn't it?

> Guess we will remove all of these casts in the near future?

I still want to clean up the code. Note that working with non-`int` numbers is always a bit of a pain, though.

**Comment by 304NotModified** [comment]
> If I'm not mistaken, Hole struct is 12 bytes on 32bits, 20 on 64bits

Yes you're right. FYI, the recommendation is max 16 bytes: https://msdn.microsoft.com/en-us/library/ms229017.aspx

> I don't plan to copy ours around much. For rendering I currently have one copy in mind, hopefully I can make it 0, but without ref-returns (C# 7) I'm not sure if it's possible.

I think the list should be non-public then, as we now can consume the holes and copy it by passing to another method. 

> To be honest, I am no expert about CLS-compliance as I never cared about this in my projects before.

CLS-compliance is still important. In summary, we can't use unsigned types in public contracts. So my proposal to make them internal/private. If we need them publicly, we should encapsulate them with a signed type. (preferable readonly)

**Comment by 304NotModified** [comment]
anyway, I think I can easily adapt #17 to this :)

**Comment by jods4** [comment]
Yes, I think most of the currently public API surface should be at least `internal` or even `private`. But at the moment it's convenient for the tests.

The `struct` are key to best perf here. With my last commit, I am never copying the `Hole` struct out of its array. Basically, rendering will read them in-place. For that usage they are perfect, no matter how large.

I think this can now be merged. So we can proceed by merging the rendering tests and impl. into this structure.

**Comment by 304NotModified** [comment]
> Yes, I think most of the currently public API surface should be at least internal or even private. But at the moment it's convenient for the tests.

Then internal is fine and the usage of `[assembly: InternalsVisibleTo()]` :)

Why did we move from `List` or array? Is that really needed for the performance?

**Comment by jods4** [comment]
> Why did we move from List or array? Is that really needed for the performance? 

Yes, arrays are special. They are the only collection in the BCL that currently allows you to access structs without copying them. `List<Hole>.this[i]` will always return you a copy. 
Since this is going to be private, I think arrays are fine.

**Comment by 304NotModified** [comment]
I like your comments in the file, but why aren't those XML comments?

**Comment by jods4** [comment]
@304NotModified Habit, I never use XML comments unless I want to produce help files for public APIs.
But if you prefer XML comments I have nothing against them.

**Comment by 304NotModified** [comment]
We prefer XML docs everywhere, as it is also visible in VS when calling the methods.
Also with merging, you get earlier an error when the merge of the docs was failed. 
Last but not least, when moving methods with refactor tools, it's nice that the docs will also move :)

If you move them to XML docs, it's ready to merge :)

**Comment by 304NotModified** [comment]
:+1:

**Pull request merged**